### DISCUSSION TITLE
How to use CursoredPageable on a Controller?

### DISCUSSION THREAD

**Discussion by radovanradic** [micronaut-projects/micronaut-data#3148] [General]
### Discussed in https://github.com/micronaut-projects/micronaut-data/discussions/3144

<div type='discussions-op-text'>

<sup>Originally posted by **rafaelrenanpacheco** September 24, 2024</sup>
I'm trying to use cursor on a `@Get` endpoint using the following code:

```java
@Get
public Page<ItemResponse> getItems(CursoredPageable page) {
    Page<Item> items = itemService.getItems(page);

    return itemMapper.toItemResponse(items);
}
```

That eventually calls this repository:

```java
@JdbcRepository(dialect = Dialect.POSTGRES)
public interface ItemRepository extends PageableRepository<Item, UUID> {

    CursoredPage<Item> findAll(CursoredPageable page);
}
```

The problem is that the request is refused because Micronaut wants a CursoredPageable that was not specified in the request. This is the output:

```json
{
  "links": {
    "self": [
      {
        "href": "/items?size=2&cursor=123",
        "templated": false,
        "type": null,
        "deprecation": null,
        "profile": null,
        "name": null,
        "title": null,
        "hreflang": null
      }
    ]
  },
  "embedded": {
    "errors": [
      {
        "links": {},
        "embedded": {},
        "message": "Required argument [CursoredPageable page] not specified",
        "logref": null,
        "path": "/page"
      }
    ]
  },
  "message": "Bad Request",
  "logref": null,
  "path": null
}
```

I've tried to use only Pageable, which works when sending the request but there is no cursor data after that. I also tried some query brute force trying to guess which query param the cursored is expecting, but without any success. The docs about cursor don't say much about using the CursoredPageable on a controller for http requests. Am I missing something trivial here?</div>

**Comment by radovanradic** [comment]
I think this is a bug, there is no mechanism to bind `CursoredPageable` from the request.

**Comment by dstepanov** [comment]
The problem is that persisting a cursor solution can be very opinionated. You might want to do hash64, include the query parameters, or maybe do JSON; there are many options.

**Reply by graemerocher** [reply]
we should provide a pluggable extension mechanism

**Reply by rafaelrenanpacheco** [reply]
I've worked on a project that used base64 to encode the cursor data on the query param. The cursor response from the backend had properties like:
- hash to the first page
- hash to the last page
- hash to next page (if there is any)
- hash to previous page (is there is any)

So the frontend could just follow those hashes without thinking too much.
